vermouth/tests/test_map_input.py

```diff
@@ -226,21 +237,97 @@ def test_read_mapping(case):
 2 X1 Y U
     """,  # Multiple difinitions for the same atom
     """
-no initial context
-    """,
-    """
-[ molecule ]
 [ atoms ]
 0 A B
     """,  # no molecule name
+    """
+Just a pile of garbage.
+Clearly, this is not a mapping file.
+Not even a partial one.
+    """,
+    # If a file contains two consecutive [ molecule ] lines, it means the first
+    # of the molecules is empty.
+    """
+[ molecule ]
+    """,
 ))
-
 def test_read_mapping_errors(content):
     """
     Test that syntax error are caught when reading a mapping.
     """
     with pytest.raises(IOError):
-        vermouth.map_input.read_mapping(content.split('\n'))
+        vermouth.map_input._read_mapping_partial(content.split('\n'), 1)  # pylint: disable=protected-access
+
+
+@pytest.mark.parametrize(
+    'case',
+    [SYSTEM_BASIC, SYSTEM_SHARED, SYSTEM_EXTRA, SYSTEM_NULL, SYSTEM_FROM_TO, SYSTEM_IGNORE]
+)
+def test_read_mapping_file(case):
+    """
+    Test that regular mapping files are read as expected.
+    """
+    reference = collections.defaultdict(lambda: collections.defaultdict(dict))
+    for from_ff, to_ff in itertools.product(case.from_ff, case.to_ff):
+        reference[from_ff][to_ff][case.name] = (
+            case.mapping, case.weights, case.extra
+        )
+    reference = vermouth.map_input._default_to_dict(reference)  # pylint: disable=protected-access
+
+    mappings = vermouth.map_input.read_mapping_file(
+        ['[ molecule ]'] + case.string.split('\n')
+    )
+
+    assert mappings == reference
+
+
+@pytest.fixture
+def reference_multi():
+    """
+    Build a reference file and mapping collection with multiple molecules.
+    """
+    content = textwrap.dedent("""
+        ; Some line before the first molecule.
+        ; Just because.
+
+        [ molecule ]
+        dummy_0
+
+        [ atoms ]
+        0 X1 A
+        1 X2 B
+
+        ; Some mess between two molecules.
+
+        [ molecule ]
+        dummy_1
+
+        [ atoms ]
+        0 X2 C
+        1 X3 D
+    """).split('\n')
```

Comment:
Comments after atom lines, comments in between atom lines, comments in the [molecule] section. Especially:
```
[ molecule ]
; Look, it's a shiny!
dummy_0
```